Ruby in WebVTT are not rendered correctly

danae404 opened the pull request:
#### def69bfc9d9e2333a4c6bc5a254169754f0f4116
<pre>
Ruby in WebVTT are not rendered correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=262064">https://bugs.webkit.org/show_bug.cgi?id=262064</a>
rdar://116011365

Reviewed by Jer Noble and Eric Carlson.

This patch creates a new class WebVTTRubyElement, a subclass of WebVTTElementImpl
And RubyElement. Now, each ruby tag in a VTT is associated with a WebVTTRubyElement,
Which has access to RubyElement::createElementRenderer() to render it appropriately.

Added a layout test to check that ruby elements in WebVTT cues are positioned correctly.

* LayoutTests/media/track/captions-webvtt/ruby.vtt: Added.
* LayoutTests/media/track/webvtt-ruby.html: Added.
* LayoutTests/media/track/webvtt-ruby-expected.txt: Added.
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesLangPseudoClass):
(WebCore::matchesFutureCuePseudoClass):
(WebCore::matchesPastCuePseudoClass):
* Source/WebCore/dom/Node.h:
(WebCore::Node::isWebVTTRubyElement const):
(WebCore::Node::isWebVTTRubyTextElement const):
* Source/WebCore/html/RubyElement.h:
(): Deleted.
* Source/WebCore/html/RubyTextElement.h:
(): Deleted.
* Source/WebCore/html/track/VTTCue.cpp:
(WebCore::VTTCue::markFutureAndPastNodes):
* Source/WebCore/html/track/WebVTTElement.cpp:
(WebCore::WebVTTElement::WebVTTElement):
(WebCore::WebVTTElementImpl::create):
(WebCore::WebVTTElementImpl::cloneElementWithoutAttributesAndChildren):
(WebCore::WebVTTElementImpl::createEquivalentHTMLElement):
(WebCore::WebVTTElement::create): Deleted.
(WebCore::WebVTTElement::cloneElementWithoutAttributesAndChildren): Deleted.
(WebCore::WebVTTElement::createEquivalentHTMLElement): Deleted.
* Source/WebCore/html/track/WebVTTElement.h:
(WebCore::WebVTTElementImpl::WebVTTElementImpl):
(isType):
* Source/WebCore/html/track/WebVTTParser.cpp:
(WebCore::WebVTTTreeBuilder::currentType const):
(WebCore::WebVTTTreeBuilder::buildFromString):
(WebCore::WebVTTTreeBuilder::constructTreeFromToken):
* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::captionsStyleSheetOverride const):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::collectMatchingShadowPseudoElementRules):

Canonical link: <a href="https://commits.webkit.org/268746@main">https://commits.webkit.org/268746@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4b549de7afc0e4f9cbe8df68cca21583db882e0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20521 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20945 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21593 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22417 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19143 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20754 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24202 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21122 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20536 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20740 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20601 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17841 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23274 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17769 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18646 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24935 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18846 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18822 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22874 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19413 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16492 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18628 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/18482 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4934 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22967 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19233 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on LayoutTests/media/track/webvtt-ruby-horizontal-expected.html:
```diff
@@ -0,0 +1,58 @@
+    requirePixelDump = true;
+    </script>
+    <script src=../media-file.js></script>
+    <script src=../video-test.js></script>
```
nice test!

danae404 replied:
Lol thank you 🤣

rr-codes commented on Source/WebCore/css/SelectorCheckerTestFunctions.h:
```diff
@@ -469,12 +473,24 @@ ALWAYS_INLINE bool matchesPictureInPicturePseudoClass(const Element& element)
 {
-    return is<WebVTTElement>(element) && !downcast<WebVTTElement>(element).isPastNode();
+    if (is<WebVTTElement>(element))
+        return !downcast<WebVTTElement>(element).isPastNode();
```
```suggestion
    if (auto* webVTTElement = dynamicDowncast<WebVTTElement>(element)
        return !webVTTElement->isPastNode();
```
I think the preferred, modern way of doing this is this, but I could be wrong.

jernoble replied:
Yep, I think @rr-codes is correct here. `is<...>` + `downcast<...>` does two two type checks, but `dynamicDowncast<...>` plus a null check does only one.

eric-carlson commented on Source/WebCore/html/track/WebVTTElement.h:
```diff
@@ -66,28 +69,89 @@ class WebVTTElement final : public Element {
+    {
+    }
+
+    bool isWebVTTRubyElement() const override { return true; }
```
Ditto

eric-carlson commented on Source/WebCore/html/track/WebVTTElement.h:
```diff
@@ -66,28 +69,89 @@ class WebVTTElement final : public Element {
+class WebVTTElement final : public WebVTTElementImpl, public Element {
+    WTF_MAKE_ISO_ALLOCATED(WebVTTElement);
+public:
+    Ref<Element> cloneElementWithoutAttributesAndChildren(Document& document) override { return WebVTTElementImpl::cloneElementWithoutAttributesAndChildren(document); }
```
final

danae404 replied:
Does override replace final or used in conjunction with it?

rr-codes replied:
I think it's used in conjunction; `override` means the function is overriding a function in its parent, and `final` means any subclasses of this class can't override it.

danae404 replied:
We spoke in person- Eric means that final should replace override

eric-carlson commented on Source/WebCore/html/track/WebVTTElement.h:
```diff
@@ -66,28 +69,89 @@ class WebVTTElement final : public Element {
+    friend class WebVTTElementImpl;
+    WebVTTElement(WebVTTNodeType, AtomString language, Document&);
+
+    bool isWebVTTElement() const override { return true; }
```
Ditto.

eric-carlson commented on Source/WebCore/html/track/WebVTTElement.h:
```diff
@@ -66,28 +69,89 @@ class WebVTTElement final : public Element {
+class WebVTTRubyElement final : public WebVTTElementImpl, public RubyElement {
+    WTF_MAKE_ISO_ALLOCATED(WebVTTRubyElement);
+public:
+    Ref<Element> cloneElementWithoutAttributesAndChildren(Document& document) override { return WebVTTElementImpl::cloneElementWithoutAttributesAndChildren(document); }
```
Ditto

eric-carlson commented on Source/WebCore/html/track/WebVTTElement.h:
```diff
@@ -66,28 +69,89 @@ class WebVTTElement final : public Element {
+class WebVTTRubyTextElement final : public WebVTTElementImpl, public RubyTextElement {
+    WTF_MAKE_ISO_ALLOCATED(WebVTTRubyTextElement);
+public:
+    Ref<Element> cloneElementWithoutAttributesAndChildren(Document& document) override { return WebVTTElementImpl::cloneElementWithoutAttributesAndChildren(document); }
```
Ditto

eric-carlson commented on Source/WebCore/html/track/WebVTTElement.h:
```diff
@@ -66,28 +69,89 @@ class WebVTTElement final : public Element {
+    {
+    }
+
+    bool isWebVTTRubyTextElement() const override { return true; }
```
Ditto

jernoble reviewed (approved):
r=me once the change from @rr-codes is addressed.

danae404 committed:
Ruby in WebVTT are not rendered correctly
https://bugs.webkit.org/show_bug.cgi?id=262064
rdar://116011365

Reviewed by Jer Noble and Eric Carlson.

This patch creates a new class WebVTTRubyElement, a subclass of WebVTTElementImpl
And RubyElement. Now, each ruby tag in a VTT is associated with a WebVTTRubyElement,
Which has access to RubyElement::createElementRenderer() to render it appropriately.

Added a layout test to check that ruby elements in WebVTT cues are positioned correctly.

* LayoutTests/media/track/captions-webvtt/ruby.vtt: Added.
* LayoutTests/media/track/webvtt-ruby.html: Added.
* LayoutTests/media/track/webvtt-ruby-expected.txt: Added.
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesLangPseudoClass):
(WebCore::matchesFutureCuePseudoClass):
(WebCore::matchesPastCuePseudoClass):
* Source/WebCore/dom/Node.h:
(WebCore::Node::isWebVTTRubyElement const):
(WebCore::Node::isWebVTTRubyTextElement const):
* Source/WebCore/html/RubyElement.h:
(): Deleted.
* Source/WebCore/html/RubyTextElement.h:
(): Deleted.
* Source/WebCore/html/track/VTTCue.cpp:
(WebCore::VTTCue::markFutureAndPastNodes):
* Source/WebCore/html/track/WebVTTElement.cpp:
(WebCore::WebVTTElement::WebVTTElement):
(WebCore::WebVTTElementImpl::create):
(WebCore::WebVTTElementImpl::cloneElementWithoutAttributesAndChildren):
(WebCore::WebVTTElementImpl::createEquivalentHTMLElement):
(WebCore::WebVTTElement::create): Deleted.
(WebCore::WebVTTElement::cloneElementWithoutAttributesAndChildren): Deleted.
(WebCore::WebVTTElement::createEquivalentHTMLElement): Deleted.
* Source/WebCore/html/track/WebVTTElement.h:
(WebCore::WebVTTElementImpl::WebVTTElementImpl):
(isType):
* Source/WebCore/html/track/WebVTTParser.cpp:
(WebCore::WebVTTTreeBuilder::currentType const):
(WebCore::WebVTTTreeBuilder::buildFromString):
(WebCore::WebVTTTreeBuilder::constructTreeFromToken):
* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::captionsStyleSheetOverride const):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::collectMatchingShadowPseudoElementRules):

Canonical link: https://commits.webkit.org/268746@main
